### ISSUE TITLE
Nested parser methods should maybe be public (cx.startNested)

### ISSUE THREAD

**Issue by Monkatraz** [lezer-parser/markdown#5]
Some flavors of Markdown add `$` (inline) and `$$` (or `$$$`) (block) for embedding TeX expressions. This is perfectly feasible to add to `lezer-markdown` using a `MarkdownConfig` extension, but it requires using private methods on the context objects.

Exposing the parser methods also opens up some support for other common extensions, like YAML front-matter, or something radical like MDX. Front-matter in particular is extremely common, and it might even benefit from a 'native' extension.

**Comment by marijnh** [comment]
You can already, with the current interface, directly parse some sub-region in your parse function and include its content in the elements you produce. This is also what the inline HTML parser does. The only reason nested parsers are supported at the block level is to make something like huge code blocks parse incrementally. I'm not sure if frontmatter blocks ever get big enough to require that.

**Comment by Monkatraz** [comment]
Other than "it would be nice to have", I do have a few examples of markup that would benefit from nested parsing:

This is what I was talking about with `$$$` blocks:
![image](https://user-images.githubusercontent.com/34875062/107751742-06a44a80-6cdb-11eb-942e-f73d4a9ab2b5.png)

This is parsed with a pretty simple Monarch grammar, but if you were to do a much more complex analysis of the markup I suspect it could get pretty heavy. You could just parse it all at once - but whenever you would change it, you would have to parse the entire thing again, practically on every keystroke. An escape hatch would be to use a fenced block syntax, but still.

Here is a real nasty one:
```
[[code type="php"]]
<?php
/* comment */
for($i=0; $i<100; $i++){
echo "number".$i."\n";
}
?>
[[/code]]
```
This is Wikidot syntax, which is actually the syntax I'm trying to replace! But if you were trying to be faithful, you couldn't really do this syntax incrementally without the nested parser methods. You also couldn't just make another grammar - Wikidot syntax has similar rules to Markdown and is equally as awful to parse.

With inline parsing - does the public interface let you embed a tree or tree nodes somehow? I see that `HTMLTag` uses `TreeElement`, but that class isn't public.

**Comment by Monkatraz** [comment]
Okay, so I've implemented nested parsing (synchronous only) and I did (have to?) use some private bits. I also used some other private bits unrelated to nested parsing, which I'll also talk about here just to keep the conversation consolidated.
```ts
// -- NAUGHTY PRIVATE API HACKERY

/** Parses the text given and returns the list of found elements.
 *  Taken from the `lezer-markdown` source. */
function parseInline(parser: MarkdownParser, text: string, offset: number): Element[] {
  let cx = new (InlineContext as any)(parser, text, offset)
  outer: for (let pos = offset; pos < cx.end;) {
    let next = cx.char(pos)
    for (let token of (parser as any).inlineParsers) if (token) {
      let result = token(cx, next, pos)
      if (result >= 0) { pos = result; continue outer }
    }
    pos++
  }
  return cx.resolveMarkers(0)
}

/** Satisfies (mostly) the interface of `Element` but takes in a `Tree` or `TreeBuffer` node instead of a type.
 *  Taken from the `lezer-markdown` source. */
class TreeElement implements Element {
  type!: number // just to make typescript happy :)
  constructor(readonly tree: Tree | TreeBuffer, readonly from: number) {}
  get to() { return this.from + this.tree.length }
  writeTo(buf: any, offset: number) {
    buf.nodes.push(this.tree)
    buf.content.push(buf.nodes.length - 1, this.from + offset, this.to + offset, -1)
  }
  toTree() { return this.tree }
}

/** Sync. parses the text given and returns a list of `TreeElements`.
 *  The language provided must be in the format of a `LanguageDescription`.
 *  If the language has not been loaded, it will be async. loaded for later usage.
 *  Taken from the `lezer-markdown` source's `HTMLTag` syntax. */
function parseNested(lang: LanguageDescription, offset: number, text: string): Element[] {
  if (!lang.support) { lang.load(); return [] }
  const parse = lang.support.language.parser.startParse(stringInput(text), 0, {})
  let tree: Tree | null = null
  while (!(tree = parse.advance())) {}
  return tree.children.map((ch, i) => new TreeElement(ch, offset + tree!.positions[i]))
}
```
In that chunk of code is:
- `parseInline` to get Markdown inline parsing in contexts other than blocks
- `TreeElement` to get `Tree` based `Element`s
- `parseNested` to parse a bit of text with a grammar and output it in a `lezer-markdown` compatible format (`TreeElement`)

Finally, there is one last private bit that I used, which is to get the position of an opening delimiter after resolving it:
```ts
const idx = cx.findOpeningDelimiter(delimiter)
if (idx !== null) {
  const start = (cx as any).parts[idx].from
  cx.takeContent(idx)
  // -snip-
}
```

If you wish to see the full context of this code, you can find it [here.](https://github.com/Monkatraz/mainframe/blob/master/packages/cm6/lib/lang.ts) If you want to see what this code is actually for, you can see my editor [here.](https://mainframe.vercel.app/edit)

I don't see any other way of doing anything those chunks of code are doing with the public interface. I could be wrong, though.

**Comment by marijnh** [comment]
> parseInline to get Markdown inline parsing in contexts other than blocks

Could you say a bit more about what you are doing when you need this?

I'll look into exporting the nested parsing and tree-node stuff.

**Comment by Monkatraz** [comment]
> Could you say a bit more about what you are doing when you need this?

Sure!

I've implemented a syntax called [CriticMarkup](http://criticmarkup.com/spec.php), which has a substitution syntax that looks like this:
```
{~~Replace this,~>With this.~~}
```
I want it to be styled like this:
![image](https://user-images.githubusercontent.com/34875062/108096235-fc1ce480-703d-11eb-91b7-c507db3dc2f0.png)

Now, because I want to be able to quickly prototype and easily construct my extensions (because I may need to add _a lot_ of new syntax, such as importable and exportable Markdown components), I've made a library of functions that automate a lot of the work for you. The implementation of critic substitution looks like this:
```ts
chain({ name: 'CriticSubstitution', predicate: '{~~',
  chain: [
    [/{~~/, '@mark'],
    [/.*?/, '@wrap Delete'],
    [/~>/,  'Arrow'],
    [/.*?/, '@wrap Insert'],
    [/~~}/, '@mark']
  ],
  tags: {
    '@mark': t.changed,
    'Delete': t.deleted,
    'Arrow': t.changed,
    'Insert': t.inserted
  }
}),

// fun fact: the regex is actually compiled into a single, optimized 'grouped' regex that is matched all at once.
// the match would be something like this:
//  ({~~) (Replace this,) (~>) (With this.) (~~})
```

The bit that required an inline parser for `InlineContext` was those `@wrap Delete` and `@wrap Insert` bits. `@wrap` tells the chain function to inline parse everything matched by that group and, optionally, further wrap those parsed elements in a node. The reason why the inline parsing / wrapping was necessary is that CriticMarkup is supposed to be for, well, criticizing work - so this syntax is likely to be added _around_ already existing markup and thus the markup that is there needs to be preserved inside of the substitution.

In order to keep the code simple and free of any quirks, I didn't want to use delimiters or some other trick, and instead just simply parse the text directly. So it takes the matched text, parses the elements out of it, and inserts it into the list of elements to be added by a successfully matched `chain` syntax. It works really well and I'm very happy with how it works.

The `parseInline` function could be optionally replaced with something like `type parse = (from: number, to: number) => Element[]` in the `InlineContext` interface.
